healthz/README.md

```diff
@@ -7,41 +14,270 @@ in distributed systems.
+persisted across restarts of the system or its hardware and software
+components, and they are removed only for resource management reasons.
+
+Whilst the system may initiate health checks itself, these should be safe to
+perform operations that do not impact the device's functionality. Expensive
+checks (e.g., pausing protocols, or dumping internal database state) that are
+potentially service impacting should require use of the `Check` RPC.
+
+`Healthz` works in conjunction with telemetry streamed via gNMI, OpenConfig
+paths for a specific component are streamed to indicate when components become
+unhealthy, allowing the receiving system to determine that further inspection
+of the component's health is required. The following paths (defined in the
+`openconfig-platform` model) are used:
+
+```
+component/healthz/state/status          (enumerated value)
+component/healthz/state/last-unhealthy  (timestamp)
+component/healthz/state/unhealthy-count (int64)
+```
+
+### Choice of Streaming for Artifacts
+
+[Background comment](https://github.com/openconfig/gnoi/pull/65#issuecomment-1090547841)
+
+Originally, the gNOI Healthz endpoint defined single transactional API for
+getting the "healthz" status of a component.  As discussed in the above issue,
+due to large data volume or to an interactive debug output there is a
+requirement that it support the ability to stream a status from the device.
+
+To ensure that this can be done using a secure channel with the relevant ACL
+and encryption, this streaming is implemented within the `Healthz` service.
+This approach further allows the concept of a historical set of collected
+artifacts to be reported by the device.
+
+### Collection of Artifacts
+
+As described above, by default health checks that are performed automatically
+by the device SHOULD NOT be performance impacting. The `Check` RPC is used to
+ensure the creation of artifacts that are expensive to collect.
+
+Thus -- a system relying on `healthz` for reporting of device status, needs
+to consider the following workflow.
 
-* The service definition provides a Get method which is correlated with a
-  /component in the OC tree. The Get method is responsible for getting the
-  current status of the requested component path. This will cause the system
-  to fetch the current state of component this may entail running specific
-  validations and gathering any relevant status of the component to reflect
-  the "health check"
+1. A component of the system becomes unhealthy. The device MUST generate
+   a health state via telemetry that indicates that the `healthz/state/status`
+   of the component has transitioned to `UNHEALTHY`. It is left for the device
+   implementor to determine which artifacts are generated automatically by the
+   system immediately. The artifact collection SHOULD NOT impact the performance
+   of the system.
+2. The responding client system observes the transition to `UNHEALTHY` and can
```

Comment:
this is an interesting idea
We would need to add an attribute which could signal there is "additional" work to be done to generate the artifacts 

It is easy to say "show tech" at the chassis is expensive but really we should never be getting "chassis" level checks we should really be only getting the needed artifacts for specific events

You do raise a reasonable use case though - let's say a system database reports an error - instead of pausing the system which could be service impacting to dumb the db- you could mark the event get the non service impacting data then leave a mark on the event that says - if you want more you can call check(event id) and it will go generate the other data and link to id which could be done after the device is drained or otherwise safe to get more data... That seems reasonable not sure it is "required" as we don't have a use case exactly of that

Reply:
Agreed. I've added the `event_id` to the `CheckRequest` to indicate that artifacts should be collected for a previous event, and made the clarifications that we discussed.

@hellt  PTAL at this too.

Reply:
@robshakir where did you add the text/field about it? I can't seem to find new commits since [92516ec](https://github.com/openconfig/gnoi/pull/106/commits/92516ecaa5bea9611cda4c21d6a7f3166b66dd31)

Reply:
I'm sorry -- this was my fault. I'd not spotted that the push needed a merge. PTAL at https://github.com/openconfig/gnoi/pull/106/commits/8cd01e1000a03f86f4dbfec6e9fc40a0c0700d6c